core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala

```diff
@@ -120,6 +125,45 @@ class KRaftQuorumImplementation(
     CoreUtils.swallow(raftManager.shutdown(), log)
     CoreUtils.swallow(controllerServer.shutdown(), log)
   }
+
+  def restartController(deleteTopicEnabled: Boolean): Unit = {
+    val prevPort = controllerQuorumVotersFuture.get().get(1000).
+      asInstanceOf[InetAddressSpec].address.getPort
+    log.info("Restarting the KRaft-based controller")
+    controllerServer.shutdown()
+
+    val props = new Properties()
+    props.putAll(controllerServer.config.originals)
+    // explicitly set the port of any addresses, as we want to re-use the port that
+    // was derived on startup (through the use of port "0") in order to have
+    // existing clients re-connect gracefully
+    props.setProperty(KafkaConfig.ListenersProp, s"CONTROLLER://localhost:$prevPort")
+    props.setProperty(KafkaConfig.QuorumVotersProp, s"1000@localhost:$prevPort")
+    log.info(s"Setting KRaft-based controller port to $prevPort as part of restart")
+    if (!deleteTopicEnabled) {
+      props.setProperty(KafkaConfig.DeleteTopicEnableProp, "false")
+    }
+    val config = new KafkaConfig(props)
+
+    val newControllerServer = new ControllerServer(
```

Comment:
it would be nice if we could figure out how to reuse some of the logic in `newKRaftQuorum`.